### DISCUSSION TITLE
How to secure mKCP transport?

### DISCUSSION THREAD

**Discussion by zonescape** [XTLS/Xray-core#3313] [Q&A] [answered]
From [the documentation](https://xtls.github.io/en/config/transports/mkcp.html) on mKCP transport:

> `seed`: string
> 
> An optional obfuscation seed is used to obfuscate traffic data using the `AES-128-GCM` algorithm...
> 
> This obfuscation mechanism cannot ensure the security of the content...

It states that `seed` cannot ensure the security of the content but [mKCP usage example](https://github.com/XTLS/Xray-examples/tree/ecefc32120d6e1bd778c6becb8f84e06b9180ac9/VLESS-mKCPSeed) doesn't contain any additional security layer over mKCP. So how to use mKCP securely? I also don't understand why the `AES-128-GCM` cipher called "obfuscation" in the aforementioned docs.

**Comment by chise0713** [comment]
Just use Vmess inside mKCP, it can provide a secure encryption for your connection.

**Reply by zonescape** [reply]
I read an article about Xray where VMess called "obsolete". So I think it would be better to try TLS first. Is this setup good enough?

client:
```json
"outbounds": [
    {
        "protocol": "vless",
        "streamSettings": {
            "network": "kcp",
            "kcpSettings": {...},
            "security": "tls",
            "tlsSettings": {
                "serverName": "zonescape.proxy",
                "certificates": [
                    {
                        "certificate": [...],
                        "usage": "verify"
                    }
                ]
            }
        }
    }
]
```
server:
```json
"inbounds": [
    {
        "protocol": "vless",
        "streamSettings": {
            "network": "kcp",
            "kcpSettings": {...},
            "security": "tls",
            "tlsSettings": {
                "certificates": [
                    {
                        "certificate": [...],
                        "key": [...]
                    }
                ]
            }
        }
    }
]
```
I use this command to generate certificate/key pair:
```bash
xray tls cert -domain zonescape.proxy -expire 87600h
```

**Reply by chise0713** [reply]
~~mKCP seems not compatible with TLS~~, and VMess is safe enough if you don't leak your uuid.

Edit:
I looked into the source code, mKCP is compatible with TLS, my mistake.
And if you want forward security, TLS is always a good option.

**Comment by zonescape** [comment, marked as answer]
I ended up with these configs. They are the same as in the [examples](https://github.com/XTLS/Xray-examples/tree/ecefc32120d6e1bd778c6becb8f84e06b9180ac9/VLESS-mKCPSeed), but with TLS. Hope it is secure enough now.
<details><summary>client.json</summary>
<p>

```json
{
    "log": {
        "loglevel": "warning"
    },
    "inbounds": [
        {
            "port": 1080,
            "listen": "127.0.0.1",
            "protocol": "socks",
            "settings": {
                "udp": true
            }
        }
    ],
    "outbounds": [
        {
            "protocol": "vless",
            "settings": {
                "vnext": [
                    {
                        "address": "{{ host }}",
                        "port": {{ port }},
                        "users": [
                            {
                                "id": "{{ uuid }}",
                                "encryption": "none"
                            }
                        ]
                    }
                ]
            },
            "streamSettings": {
                "network": "kcp",
                "kcpSettings": {
                    "seed": "{{ seed }}"
                },
                "security": "tls",
                "tlsSettings": {
                    "serverName": "some.domain",
                    "certificates": [
                        {
                            "certificate": [...],
                            "usage": "verify"
                        }
                    ]
                }
            }
        }
    ]
}
```
</p>
</details>

<details><summary>server.json</summary>
<p>

```json
{
    "log": {
        "loglevel": "warning"
    },
    "inbounds": [
        {
            "protocol": "vless",
            "port": {{ port }}, 
            "settings": {
                "decryption":"none",
                "clients": [
                    {
                        "id": "{{ uuid }}"
                    }
                ]
            },
            "streamSettings": {
                "network": "kcp",
                "kcpSettings": {
                    "seed": "{{ seed }}"
                },
                "security": "tls",
                "tlsSettings": {
                    "certificates": [
                        {
                            "certificate": [...],
                            "key": [...]
                        }
                    ]
                }
            }
        }
    ],
    "outbounds": [
        {
            "protocol": "freedom"
        }
    ]
}
```
</p>
</details>

Certificate/key pair is generated using this command (certificate will expire in 10 years):
```bash
xray tls cert -domain some.domain -expire 87600h
```

**Reply by donqyu** [reply]
Hi thanks for your great sharing. I finally find a safer way to keal with kcp here.
Is there any update for this setup since it has been 10 month since you post it? Is it still safely in use?

**Reply by zonescape** [reply]
I used this config for a while but stopped due to low connection speed. I have 100 Mb/s link and can't get more then 14 Mb/s from this config. There is also strange effect when the speed drops abruptly. I tried to use [kcptun](https://github.com/xtaci/kcptun) directly (without Xray) with OpenVPN (TCP mode) and got the same result. Low connection speed with abrupt decreases. One can suspect these abrupt decreases related to censorship but I don't think so. I tried other UDP proxies and all of them utilize full bandwidth of my internet connection without any quirks.
